File: src/main/java/com/blockninja/counterclockwise/ships/BeltStorageForceInducer.java
package com.blockninja.counterclockwise.ships;

import com.blockninja.counterclockwise.jackson.BlockPosKeySerializer;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import com.simibubi.create.content.kinetics.belt.BeltPart;
import com.simibubi.create.content.kinetics.belt.BeltSlope;
import net.minecraft.core.BlockPos;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.world.level.Level;
import org.jetbrains.annotations.NotNull;
import org.valkyrienskies.core.api.ships.*;
import org.valkyrienskies.core.api.world.PhysLevel;
import org.valkyrienskies.mod.common.VSGameUtilsKt;

import java.util.concurrent.ConcurrentHashMap;

public final class BeltStorageForceInducer implements ShipPhysicsListener {

    public ConcurrentHashMap<Long, BeltData> beltLocations = new ConcurrentHashMap<>();

    @Override
    public void physTick(@NotNull PhysShip physShip, @NotNull PhysLevel physLevel) {
        // Do fuck all
    }

    public static BeltStorageForceInducer getOrCreate(LoadedServerShip ship) {
        BeltStorageForceInducer attachment = ship.getAttachment(BeltStorageForceInducer.class);
        if (attachment == null) {
            attachment = new BeltStorageForceInducer();
            ship.setAttachment(attachment);
        }
        return attachment;
    }

    public static BeltStorageForceInducer get(Level level, BlockPos pos) {
        ServerLevel serverLevel = (ServerLevel) level;

        LoadedServerShip ship = VSGameUtilsKt.getShipObjectManagingPos(serverLevel, pos);

        return ship != null ? getOrCreate(ship) : null;
    }

    public record BeltData(BeltSlope slop, BeltPart part, float speed) {}
}

File: src/main/kotlin/com/blockninja/counterclockwise/ships/BeltStorageForceInducer.kt
package com.blockninja.counterclockwise.ships

import com.simibubi.create.content.kinetics.belt.BeltPart
import com.simibubi.create.content.kinetics.belt.BeltSlope
import net.minecraft.core.BlockPos
import net.minecraft.core.Direction
import net.minecraft.server.level.ServerLevel
import net.minecraft.world.level.Level
import org.valkyrienskies.core.api.ships.LoadedServerShip
import org.valkyrienskies.core.api.ships.PhysShip
import org.valkyrienskies.core.api.ships.ShipPhysicsListener
import org.valkyrienskies.core.api.util.GameTickOnly
import org.valkyrienskies.core.api.util.PhysTickOnly
import org.valkyrienskies.core.api.world.PhysLevel
import org.valkyrienskies.mod.common.getShipObjectManagingPos
import java.util.concurrent.ConcurrentHashMap

class BeltStorageForceInducer : ShipPhysicsListener {
    @JvmField
    var beltLocations: ConcurrentHashMap<Long, BeltData> = ConcurrentHashMap<Long, BeltData>()

    @OptIn(PhysTickOnly::class)
    override fun physTick(physShip: PhysShip, physLevel: PhysLevel) {
        // Do fuck all
    }

    @JvmRecord
    data class BeltData(val slop: BeltSlope, val part: BeltPart, val speed: Float, val direction: Direction)

    companion object {
        @OptIn(GameTickOnly::class)
        @JvmStatic
        fun getOrCreate(ship: LoadedServerShip): BeltStorageForceInducer {
            var attachment = ship.getAttachment(BeltStorageForceInducer::class.java)
            if (attachment == null) {
                attachment = BeltStorageForceInducer()
                ship.setAttachment(attachment)
            }
            return attachment
        }

        @OptIn(GameTickOnly::class)
        fun get(level: Level, pos: BlockPos): BeltStorageForceInducer? {
            val serverLevel = level as ServerLevel?

            val ship = serverLevel.getShipObjectManagingPos(pos)

            return if (ship != null) getOrCreate(ship) else null
        }
    }
}